### PR TITLE
Initial Commit For api-01-local-function Objective

### LINKED ISSUE #35: api-01-local-function / local api function call
- [x] objective.json
- [x] description.md
- [x] validator.js
- [x] walkthrough.md

For more details, reference the [original objective planning doc](https://docs.google.com/document/d/1Esz69gd10vK4Btnk_pn0pmKTUgIq9f7sCrmnMVyOt5E/edit#heading=h.uoyv4k1cq39u).

## Objective outline
- This is the first step in learning how to craft the Divination spell
- Spells are crafted by passing data around? 
- Intro APIs as a local function call of `String.replaceAll()`
- Write a QuestIDE function that turns all instance of "word1" with "word2" 
  - What is our divination/API themed phrase and keywords to replace?
  - Probably we want a few different phrases as test cases to avoid hard coding

### PR DESCRIPTION
**Pull request by MarvelProgramming**
For issue #35

### CODE REVIEW

**Comment by MarvelProgramming** [comment]
As it stands, each of the objectives (i.e. this one, #57, #58, #59, and #60) are defining a `DIVINATION_API_ENDPOINT` constant in their validator. I feel as though adding this to the `scripts/config` and referencing that instead would be better overall. Open to thoughts here!

**Comment by rmkubik** [comment]
I think we'd want these objectives in the `lovelace_tower` level from the feature branch:
https://github.com/TwilioQuest/twilioquest-api-academy/tree/milestone-1-aa3/levels/lovelace_tower

**Comment by rmkubik** [comment]
General outline looks good! Using a var in scripts/config makes sense. Guessing that's in the other objective drafts though, not this one? Since it doesn't hit the backend

**Comment by MarvelProgramming** [comment]
> I think we'd want these objectives in the `lovelace_tower` level from the feature branch: https://github.com/TwilioQuest/twilioquest-api-academy/tree/milestone-1-aa3/levels/lovelace_tower

My current plan is to merge all `milestone-1-aa3` into each of the objective branches and change the PR base. Thought about rebasing, but that doesn't feel like the best way to go because the branches are already here on the remote

\*UPDATE\*: All done!

> Guessing that's in the other objective drafts though, not this one? Since it doesn't hit the backend

Exactly!

File: levels/lovelace_tower/objectives/api-01-local-function/walkthrough.md [hunk 1]
```diff
@@ -0,0 +1,4 @@
+# Helpful Resources
+
+- [MDN - String replaceAll method](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replaceAll)
```
Comment (MarvelProgramming) [inline levels/lovelace_tower/objectives/api-01-local-function/walkthrough.md:7]: Not sure what (if anything) more we wanted to do with the walkthrough for this objective. We're not really introducing anything new to the player, besides maybe the `replaceAll` method. Welcome to any ideas!

File: levels/lovelace_tower/objectives/api-01-local-function/description.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+<ul>
+  <li>Create a function called `wordSwapper`.</li>
+  <li>This function receives 3 string arguments.</li>
+  <li>Replace all instances of the "target" string in the "source" string with the "replacement" string, using the `replaceAll` function.</li>
+  <li>Return the resulting string.</li>
+  <li>Once you're done, press <em>HACK</em>.</li>
+</ul>
+</div>
+
+Before we jump into working with Web APIs let's try interacting with a function locally. For this, we'll use the `String.replaceAll` method to replace all instances of a particular string with something else.
```
Comment (joenash) [inline levels/lovelace_tower/objectives/api-01-local-function/description.md]:
```suggestion
Before we jump into working with Web APIs let's remind ourselves about local APIs provided by libraries, as functions and objects. For this, we'll use the `String.replaceAll` method to replace all instances of a particular string with something else.
```

Minor suggestion to make the "libraries expose APIs too" link explicit

File: levels/lovelace_tower/objectives/api-01-local-function/description.md [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+```js
+const myString = "one two one two one two";
+const swappedString = myString.replaceAll("two", "zero");
+
+console.log(swappedString);
+```
+
+would print `one zero one zero one zero` to the console.
+
+Create a function called `wordSwapper` that takes 3 string arguments -- `sourceString`, `targetWord`, and `replacementWord`. Then replace all instances of `targetWord` in the `sourceString` with `replacementWord` and return the result.
```
Comment (joenash) [inline levels/lovelace_tower/objectives/api-01-local-function/description.md]: `that takes 3 string arguments -- `sourceString`, `targetWord`, and `replacementWord`. `
Would these 3 elements be better as a bulleted list?

File: levels/lovelace_tower/objectives/api-01-local-function/walkthrough.md (hunk 1 above)
Comment (joenash) [reply levels/lovelace_tower/objectives/api-01-local-function/walkthrough.md:7]: Maybe suggest they revisit the Turing initiation in part 1? That covers the local API topic

**Review by joenash — APPROVED** [review]
A couple of minor suggestions but nothing breaking! Looking great

File: levels/lovelace_tower/objectives/api-01-local-function/description.md (hunk 2 above)
Comment (MarvelProgramming) [reply levels/lovelace_tower/objectives/api-01-local-function/description.md]: I like this! It really hits home on the idea that APIs extend beyond the Web!

File: levels/lovelace_tower/objectives/api-01-local-function/description.md (hunk 3 above)
Comment (MarvelProgramming) [reply levels/lovelace_tower/objectives/api-01-local-function/description.md]: I think so!

File: levels/lovelace_tower/objectives/api-01-local-function/walkthrough.md (hunk 1 above)
Comment (MarvelProgramming) [reply levels/lovelace_tower/objectives/api-01-local-function/walkthrough.md:7]: ~~Is there a specific part of the Turing initiation you're referring to? I was having trouble finding this while playing through the scenario. There is the tome at the beginning of the `lovelace_tower/default` map that links to resources for APIs, if that's what you are pointing to!~~

\*UPDATE\*: Scratch that, had a bit of a brain fart

**Comment by MarvelProgramming** [comment]
From what I recall, we wanted to tie this objective into the corrupted inscriptions the player later stumbles upon in the library. How did we want to tackle this without spoiling the story parts of the library/api-05? This may not be something we're worried about, but thought I'd check in before changing things further @rmkubik @nmadadair

**Review by rmkubik — APPROVED** [review]
We talked about swapping the replacement strings to be more thematic. Since we're passing in the strings during validation though, I'm less worried about it.

**Comment by nmadadair** [comment]
In terms of the narrativization of this objective, we could add something pretty simple at the very beginning of the description that acknowledges that this objective is the very first challenge of the House Gauntlet. Something along the lines of: 

> _Welcome, Gauntleter, to the first of many challenges in the Arcane Academy's annual House Gauntlet. As you progress through Lovelace Tower, you will be tested on your ability to apply your knowledge of APIs. For this first challenge, the House Gauntlet asks you to explore how a local function API can manipulate the written word, as befits House Lovelace._

My idea here is just to call out that we're able to use APIs to change text and language, and thematically House Lovelace is the house of documentation and books and thus text/language, etc. To your question above, @marvjohnson, about foreshadowing the Fredric corruption objective without giving it away, we could add a note at the end of the description  about how: 

> _Local function APIs like `swapStrings()`, though simple, have the power to alter not just individual sentences, but entire tomes. A Gauntleter might imagine the chaos if one were to swap every " " in a House Lovelace volume for a "" -- how easily important documents could be corrupted -- or repaired -- through such simple APIs!_

File: levels/lovelace_tower/objectives/api-01-local-function/walkthrough.md (hunk 1 above)
Comment (nmadadair) [reply levels/lovelace_tower/objectives/api-01-local-function/walkthrough.md:7]: Tomb -> tome

**Comment by MarvelProgramming** [comment]
> We talked about swapping the replacement strings to be more thematic. Since we're passing in the strings during validation though, I'm less worried about it.

Went in and made some changes such that the examples mesh with the story a bit better! Also @nmadadair, I went in and implemented the changes you suggested!

**Pull request merged**